File: game-service/src/main/java/com/insecurebank/game/config/DatabaseConfig.java
package com.insecurebank.game.config;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

import javax.sql.DataSource;
import java.sql.SQLException;

@Component
public class DatabaseConfig {

    private final DataSource gameDataSource;
    private final DataSource bankDataSource;

    public DatabaseConfig(@Qualifier("gameDataSource") DataSource gameDataSource,
                           @Qualifier("bankDataSource") DataSource bankDataSource) {
        this.gameDataSource = gameDataSource;
        this.bankDataSource = bankDataSource;
    }

    @EventListener(ApplicationReadyEvent.class)
    public void verifyDatabaseConnections() {
        verify("game_mfa", gameDataSource);
        verify("auth_db (shared bank credentials)", bankDataSource);
    }

    private void verify(String label, DataSource dataSource) {
        int retryCount = 0;
        int maxRetries = 30;

        System.out.println("Verifying PostgreSQL connection for schema: " + label);

        while (retryCount < maxRetries) {
            try (var connection = dataSource.getConnection();
                 var statement = connection.createStatement();
                 var resultSet = statement.executeQuery("SELECT 1")) {
                if (resultSet.next()) {
                    System.out.println("Connection verified for schema: " + label);
                    return;
                }
            } catch (SQLException e) {
                retryCount++;
                System.out.println("Database connection failed for " + label + " (attempt " + retryCount + "/" + maxRetries + "): " + e.getMessage());
                if (retryCount < maxRetries) {
                    try {
                        Thread.sleep(5000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        return;
                    }
                } else {
                    throw new RuntimeException("Failed to connect to PostgreSQL schema " + label + " after " + maxRetries + " attempts", e);
                }
            }
        }
    }
}

File: auth-service/src/main/kotlin/com/insecurebank/auth/config/DatabaseConfig.kt
package com.insecurebank.auth.config

import org.springframework.boot.context.event.ApplicationReadyEvent
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component
import javax.sql.DataSource
import java.sql.SQLException

@Component
class DatabaseConfig(
    private val dataSource: DataSource
) {
    
    @EventListener(ApplicationReadyEvent::class)
    fun verifyDatabaseConnection() {
        var retryCount = 0
        val maxRetries = 30
        
        println("🔍 Verifying PostgreSQL database connection for auth service...")
        
        while (retryCount < maxRetries) {
            try {
                dataSource.connection.use { connection ->
                    connection.createStatement().use { statement ->
                        statement.executeQuery("SELECT 1").use { resultSet ->
                            if (resultSet.next()) {
                                println("✅ Auth service PostgreSQL database connection verified")
                                println("📊 Connected to schema: auth_db")
                                return
                            }
                        }
                    }
                }
            } catch (e: SQLException) {
                retryCount++
                println("❌ Database connection failed (attempt $retryCount/$maxRetries): ${e.message}")
                
                if (retryCount < maxRetries) {
                    println("🔄 Retrying database connection in 5 seconds...")
                    Thread.sleep(5000)
                } else {
                    println("❌ Max database connection retries reached for auth service")
                    throw RuntimeException("Failed to connect to PostgreSQL database after $maxRetries attempts", e)
                }
            }
        }
    }
}